### PR TITLE
feat: refactor, add ts and wasm

### PR DESCRIPTION
**Pull request by driemworks**
- includes the wasm bindings and typescript wrapper for them (formerly etf-sdk-wasm and etf.js, respectively)
- tightly couples the tle rust library, wasm bindings, and typescript wrapper, allowing for easy maintenance of updates to the API as the library evolves, as the ts piece of the library directly depends on the rust one.
- tle.js becoms etf.js and we no longer publish the wasm-pack output to npm
    -  The build script is modified to build the wasm and freshly install dependencies before transpiling to typescript
- also includes several update to the tlock API to allow for greater flexibility:
    - introduces a StreamCipherProvider trait and AESGCMStreamCipherProvider implementation. This allows for the library to be unopinionated, allowing an implementor to specify the encryption mechanism and CSPRNG used. For example, we could implement tlock using [age](https://github.com/str4d/rage) in the near future

### CODE REVIEW

File: ts/src/tle.test.spec.ts [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    expect(mockCallback).toHaveBeenCalledWith(expect.objectContaining({
+      signaturesCompact: expect.any(Array)
+    }));
+
+    // Fast-forward another 3 seconds
+    jest.advanceTimersByTime(3000);
+    expect(mockCallback).toHaveBeenCalledTimes(3);
+    expect(mockCallback).toHaveBeenCalledWith(expect.objectContaining({
+      signaturesCompact: expect.any(Array)
+    }));
```
Comment (juangirini) [inline ts/src/tle.test.spec.ts:70]: why this is not in a loop?

File: ts/src/tle.test.spec.ts [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+  });
+
+  it('should call getPulse', async () => {
+    const etf = new Etf('wss://example.com', true);
+    await etf.init()
```
Comment (juangirini) [inline ts/src/tle.test.spec.ts:75]:
```suggestion
    const tle = new Tle('wss://example.com', true);
    await tle.init()
```

File: ts/src/tle.test.spec.ts [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+    etf.getPulse(0).then(pulse => {
```
Comment (juangirini) [inline ts/src/tle.test.spec.ts:77]:
```suggestion
    tle.getPulse(0).then(pulse => {
```

File: ts/src/tle.test.spec.ts [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    });
+  });
+
+  it('should timelock encrypt a message', async () => {
+    const etf = new Etf()
+    await etf.init(JSON.stringify(chainSpec), false)
+    const seed = 'seed';
+    const latestBlockNumber = 123;
+    const message = 'Hello, world!'
+    await etf.timelockEncrypt(new TextEncoder().encode(message), latestBlockNumber, seed).then((result) => {
```
Comment (juangirini) [inline ts/src/tle.test.spec.ts:90]:
```suggestion
    const tle = new Tle()
    await tle.init(JSON.stringify(chainSpec), false)
    const seed = 'seed';
    const latestBlockNumber = 123;
    const message = 'Hello, world!'
    await tle.timelockEncrypt(new TextEncoder().encode(message), latestBlockNumber, seed).then((result) => {
```

File: ts/src/tle.test.spec.ts [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+      let result_string = JSON.stringify(result);
+      let expected_string = JSON.stringify({
+        aes_ct: { ciphertext: [ 0 ], nonce: [ 1 ] },
+        etf_ct: 'mocked-etf-ct',
```
Comment (juangirini) [inline ts/src/tle.test.spec.ts:94]: is this correct? or should it be `tle_ct`?

File: ts/src/tle.test.spec.ts [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+      expect(result_string).toBe(expected_string);
+    });
+  })
+
+  it('should timelock decrypt a message', async () => {
+    const etf = new Etf()
+    await etf.init(JSON.stringify(chainSpec), false)
+    const blockNumber = 1;
+    const ciphertext = 'ciphertext'
+    const result = await etf.timelockDecrypt(ciphertext, blockNumber);
```
Comment (juangirini) [inline ts/src/tle.test.spec.ts:111]:
```suggestion
    const tle = new Tle()
    await tle.init(JSON.stringify(chainSpec), false)
    const blockNumber = 1;
    const ciphertext = 'ciphertext'
    const result = await tle.timelockDecrypt(ciphertext, blockNumber);
```

File: ts/src/tle.test.spec.ts [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+    expect(result).toEqual({
+      message: 'mocked-decrypted',
+      sk: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1]
+    })
+  })
+
+  it('should decrypt a message on demand if the user knows the secret', async () => {
+    const etf = new Etf()
+    await etf.init(JSON.stringify(chainSpec), false)
```
Comment (juangirini) [inline ts/src/tle.test.spec.ts:120]:
```suggestion
    const tle = new Tle()
    await tle.init(JSON.stringify(chainSpec), false)
```

File: ts/src/tle.ts [hunk 8]
```diff
@@ -0,0 +1,178 @@
+// Encryption to the Future
+// This class initializes the ETF.js SDK
```
Comment (juangirini) [inline ts/src/tle.ts:2]:
```suggestion
// Timelock Encryption
// This class initializes the tle.js SDK
```

File: ts/src/tle.ts [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+import { BN, BN_ONE, hexToString, hexToU8a } from "@polkadot/util";
+import init, { build_encoded_commitment, tle, tld, decrypt } from 'tle-wasm-wrapper'
+import hkdf from 'js-crypto-hkdf'; // for npm
+import { Pulse, Justfication } from './types'
+
+/**
+ * Encryption to the Future
+ * This class initializes the ETF.js SDK
+ */
+export class Etf {
```
Comment (juangirini) [inline ts/src/tle.ts:18]:
```suggestion
 * Timelock Encryption
 * This class initializes the tle.js SDK
 */
export class Tle {
```

File: ts/src/tle.ts [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+  public api!: ApiPromise
+  private providerMultiAddr: string
+  private readonly MAX_CALL_WEIGHT2 = new BN(1_000_000_000_000).isub(BN_ONE);
+  private readonly MAX_CALL_WEIGHT = new BN(5_000_000_000_000).isub(BN_ONE);
+  private readonly PROOFSIZE = new BN(1_000_000_000);
+  private readonly HASH = 'SHA-256';
+  private readonly HASHLENGTH = 32;
+
+  /**
+   * Constructor for the etf api
```
Comment (juangirini) [inline ts/src/tle.ts:30]:
```suggestion
   * Constructor for the tle api
```

File: ts/src/tle.ts [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+   * @param providerMultiAddr (optional): The multiaddress of an RPC node
+   * e.g. insecure local node:    ws://localhost:9944 
+   *      secure websocket (rpc): wss://etf1.idealabs.network:443
```
Comment (juangirini) [inline ts/src/tle.ts:33]: This will need an update https://github.com/ideal-lab5/tle/issues/2

File: ts/src/tle.ts [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+    providerMultiAddr?: string,
+    isProd?: boolean,
+  ) {
+    this.providerMultiAddr = providerMultiAddr
+    this.isProd = isProd
+  }
+
+  /**
+   * Connect to the chain and start etf api wrapper
+   * @param chainSpec The ETF Network (raw) chain spec
```
Comment (juangirini) [inline ts/src/tle.ts:45]:
```suggestion
   * Connect to the chain and start tle api wrapper
   * @param chainSpec The Ideal Network (raw) chain spec
```

File: ts/src/tle.ts [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+    this.api = await ApiPromise.create({
+      provider,
+      types: {
+        ...extraTypes, Pulse
+      }
+    })
+    await init();
+    await this.api.isReady
+
+    this.ibePubkey = await this.api.query.etf.roundPublic()
```
Comment (juangirini) [inline ts/src/tle.ts:69]: This will need an update

File: ts/src/tle.ts [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+  createType(typeName: string, typeData: any): any {
+    return this.api.registry.createType(typeName, typeData);
+  }
+
+  /**
+   * listens for incoming justifications and invokes the callback when new ones are streamed
+   * @param callback: a callback to handle the new justifications
+   */
+  subscribeBeacon(callback: any): void {
+    this.api.rpc.beefy.subscribeJustifications((sig) => {
```
Comment (juangirini) [inline ts/src/tle.ts:85]: Is this the real beefy or the beefy-etf?

File: ts/src/tle.ts [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+   * Timelock Encryption: Encrypt the message for the given block
+   * @param message: The message to encrypt
+   * @param blockNumber: The block number when the message unlocks
+   * @param seed: A seed to derive crypto keys
+   * @returns the ciphertext
+   */
+  timelockEncrypt(encodedMessage: Uint8Array, blockNumber: number, seed: string): Promise<any> {
+    // TODO: fine for now but should ultimately query the BABE pallet config instead
+    // let epochLength = 200;
+    // let validatorSetId = blockNumber % epochLength;
```
Comment (juangirini) [inline ts/src/tle.ts:115]:
```suggestion
```

File: ts/src/tle.ts [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+    let t = new TextEncoder();
+    let masterSecret = t.encode(seed);
+    return hkdf.compute(masterSecret, this.HASH, this.HASHLENGTH, '').then((derivedKey) => {
+      let commitment = build_encoded_commitment(blockNumber, 0);
+      // let encodedMessage = t.encode(message);
+      // let encodedMessage = message;
```
Comment (juangirini) [inline ts/src/tle.ts:121]:
```suggestion
```

File: wasm/src/lib.rs [hunk 17]
```diff
@@ -0,0 +1,485 @@
...
+// TODO: proper error handling
+/// a helper function to deserialize arkworks elements from bytes
+fn convert_from_bytes<E: CanonicalDeserialize, const N: usize>(
+	bytes: &[u8; N],
+) -> Option<E> {
+	E::deserialize_compressed(&bytes[..]).ok()
+}
+
+/// The encrypt wrapper used by the WASM blob to call tlock.rs encrypt function
+/// in etf-crypto-primitives
```
Comment (juangirini) [inline wasm/src/lib.rs:42]: does it still hold that name

File: wasm/src/lib.rs [hunk 18]
```diff
@@ -0,0 +1,485 @@
...
+	ciphertext
+		.serialize_compressed(&mut ciphertext_bytes)
+		.map_err(|_| JsError::new("ciphertext serialization has failed"))?;
+
+	serde_wasm_bindgen::to_value(&ciphertext_bytes)
+		.map_err(|_| JsError::new("could not convert ciphertext to JsValue"))
+}
+
+/// The decrypt wrapper used by the WASM blob to call tlock.rs encrypt function
+/// in etf-crypto-primitives
```
Comment (juangirini) [inline wasm/src/lib.rs:98]: same here

File: wasm/src/lib.rs [hunk 19]
```diff
@@ -0,0 +1,485 @@
...
+///   key and decrypt the message
+#[wasm_bindgen]
+pub fn tld(
+	ciphertext_js: JsValue,
+	sig_vec_js: JsValue,
+) -> Result<JsValue, JsError> {
+	let sig_conversion: Vec<u8> =
+		serde_wasm_bindgen::from_value(sig_vec_js.clone())
+			.map_err(|_| JsError::new("could not decode secret key"))?;
+	// let sig_value = sig_conversion.get(0).unwrap();
```
Comment (juangirini) [inline wasm/src/lib.rs:110]:
```suggestion
```

File: wasm/src/lib.rs [hunk 20]
```diff
@@ -0,0 +1,485 @@
...
+		serde_wasm_bindgen::from_value(ciphertext_js.clone())
+			.map_err(|_| JsError::new("could not decode ciphertext"))?;
+	let ciphertext_bytes: &[u8] = ciphertext_vec.as_slice();
+
+	let ciphertext: TLECiphertext<TinyBLS377> =
+		TLECiphertext::deserialize_compressed(ciphertext_bytes)
+			.map_err(|_| JsError::new("Could not deserialize ciphertext"))?;
+	// let decrypt_result: DecryptionResult=
+	// ciphertext.decrypt(sig_vec).map_err(|_| JsError::new("decryption has
+	// failed"))?;
```
Comment (juangirini) [inline wasm/src/lib.rs:128]:
```suggestion
```

File: wasm/src/lib.rs [hunk 21]
```diff
@@ -0,0 +1,485 @@
...
+
+//     // equivalent to msk.sign(message) fyi
+//     let sig: IBESecret<TinyBLS377> = identity.extract(msk);
+//     let sig_vec = vec![sig];
+//     let mut sig_bytes: Vec<_> = Vec::new();
+//     sig_vec.serialize_compressed(&mut sig_bytes).unwrap();
+
+//     serde_wasm_bindgen::to_value(&sig_bytes).map_err(|_|
+// JsError::new("extraction failed")) }
+
```
Comment (juangirini) [inline wasm/src/lib.rs:270]:
```suggestion
```

File: wasm/src/lib.rs [hunk 22]
```diff
@@ -0,0 +1,485 @@
...
+		let msk: <E as EngineBLS>::Scalar =
+			convert_from_bytes::<<E as EngineBLS>::Scalar, 32>(&sk.clone())
+				.unwrap();
+		let identity = Identity::new(b"", vec![identity_vec]);
+
+		let sig: E::SignatureGroup = identity.extract::<E>(msk).0;
+		let mut sig_bytes: Vec<_> = Vec::new();
+
+		if succesful_decrypt {
+			// sig_vec.serialize_compressed(&mut sig_bytes).unwrap();
```
Comment (juangirini) [inline wasm/src/lib.rs:323]:
```suggestion
```

**Review by juangirini — CHANGES_REQUESTED** [review]
I know that many of the requested changes come inherited from the source repo. But we need to clean up.
In general, we can't merge commented out code to the main branch.
There are still mentions to ETF everywhere.

There are also more things to do but I reckon they should be done as separate [issues](https://github.com/ideal-lab5/tle/issues), these are:
- code formatting
- use mocks in the tests
- update the RPCs urls

**Comment by driemworks** [comment]
> I know that many of the requested changes come inherited from the source repo. But we need to clean up.
> 
>     * Generally, we can't merge commented-out code to the main branch.
> 
>     * There are still mentions of ETF everywhere.
> 
>     * Some nits, like using capital letters to start docs comments, or add descriptions to todos
> 
> 
> There are also more things to do but I reckon they should be done as separate [issues](https://github.com/ideal-lab5/tle/issues), these are:
> 
>     * code formatting
the same rustfmt.toml from the murmur crate is used here, except I shortened line lengths to 80 instead of 100, can change it back if you prefer but I like the shorter lines more. 

>     * use mocks in the tests
What tests require mocks?

>     * update the RPCs urls
- which rpcs? there aren't any here, I think this comment may have been for a diff repo?

File: tle/src/stream_ciphers.rs [hunk 23]
```diff
@@ -0,0 +1,229 @@
...
+	/// AES-GCM decryption
+	///
+	/// * `ciphertext`: the ciphertext to decrypt
+	/// * `nonce`: the nonce used on encryption
+	/// * `key`: the key used for encryption
+	fn decrypt(ct: Self::Ciphertext, key: [u8; 32]) -> Result<Vec<u8>, Error> {
+		let cipher =
+			Aes256Gcm::new_from_slice(&key).map_err(|_| Error::InvalidKey)?;
+		// lets check the nonce... not great way to do it but ok for now
+		// TODO:get a valid nonce size as a constant
```
[the remark this answers is not in the document]
Comment (driemworks) [reply tle/src/stream_ciphers.rs]: no, it just needs to replace 12 with a const, will include in the next commit

**Comment by driemworks** [comment]
>     * Generally, we can't merge commented-out code to the main branch.

we should implement a process that enforces this.
 
>     * There are still mentions of ETF everywhere.
The majority of these mentions are in the copied TS wrapper (etf.js). I removed references in the core/wasm libs but created a new issue to address etf.js (and migrate it to tle.js or whichever name we land on: https://github.com/ideal-lab5/tle/issues/6

File: tle/Cargo.toml [hunk 24]
```diff
@@ -0,0 +1,77 @@
+[package]
+name = "tle"
```
[the remark this answers is not in the document]
Comment (driemworks) [reply tle/Cargo.toml:2]: maybe `timelock`, `timelock-rs`, or `tlock-rs`? none of those are used yet

File: ts/README.md [hunk 25]
```diff
@@ -0,0 +1,197 @@
...
+
+Note: You can connect to the test network by specifying `ws = 'wss://etf1.idealabs.network:443'`
+
+#### Smoldot
+
+To run with an in-browser light client (smoldot), the library is initalized with:
+
+```javascript
+let etf = new Etf()
+await etf.init(chainSpec)
```
[the remark this answers is not in the document]
Comment (driemworks) [reply ts/README.md:54]: will update this as part of #6

File: ts/README.md [hunk 26]
```diff
@@ -0,0 +1,197 @@
...
+**Decryption**
+
+```javascript
+let m = await etf.decrypt(ciphertext, nonce, capsule, blockNumbers)
+let message = String.fromCharCode(...m)
+```
+
+### Delayed Transactions
+
+Delayed transactions can be submitted by  using the `etf.delay` API.
```
[the remark this answers is not in the document]
Comment (driemworks) [reply ts/README.md:117]: #6

File: ts/README.md [hunk 27]
```diff
@@ -0,0 +1,197 @@
...
+
+See the [react-delayed-txs](./examples/react-delayed-txs//) example.
+
+``` javascript
+// the call to delay
+let innerCall = etf.api.tx.balances
```
[the remark this answers is not in the document]
Comment (driemworks) [reply ts/README.md:123]: #6

File: timelock/src/stream_ciphers.rs [hunk 28]
```diff
@@ -41,48 +41,61 @@ pub struct AESOutput {
 	pub nonce: Vec<u8>,
 }
 
+/// The expected length of a nonce used with AES_GCM
+const AES_GCM_NONCE_LEN: usize = 12;
+
+/// Errors that mayb be encountered with using a stream cipher
```
Comment (juangirini) [inline timelock/src/stream_ciphers.rs:47]:
```suggestion
/// Errors that may be encountered with using a stream cipher
```

File: ts/src/index.js [hunk 29]
```diff
@@ -0,0 +1 @@
+export * from './tle';
```
Comment (juangirini) [inline ts/src/index.js:1]:
```suggestion
export * from './timelock';
```

File: ts/package.json [hunk 30]
```diff
@@ -1,58 +1,33 @@
 {
-  "name": "@ideallabs/tle.js",
-  "version": "0.0.1-dev",
-  "description": "a typescript sdk for timelock encryption/decryption with the Ideal Network",
+  "name": "@ideallabs/timelock.js",
+  "version": "0.0.1",
+  "description": "A typescript interface for timelock encryption",
   "license": "GPL-3.0",
   "repository": "https://github.com/ideal-lab5/tle",
```
Comment (juangirini) [inline ts/package.json:6]:
```suggestion
  "repository": "https://github.com/ideal-lab5/timelock",
```

File: ts/src/tle.js
Comment (juangirini) [inline ts/src/tle.js:1]: Shouldn't this file be renamed to `timelock`?

File: ts/src/tle.js [hunk 31]
```diff
@@ -0,0 +1,60 @@
...
+ * @param roundNumber: The round of the protocol
+ * @param identityBuilder: Something that imlement IdentityBuilder (e.g. idealNetworkIdentityHandler)
+ * @param beaconPublicKey: The public key of the randomness beacon used
+ * @param seed: A seed to derive crypto keys
+ * @returns the ciphertext
+ */
+export async function timelockEncrypt(encodedMessage, roundNumber, identityBuilder, beaconPublicKey, seed) {
+    await init();
+    // TODO: fine for now but should ultimately query the BABE pallet config instead
+    // https://github.com/ideal-lab5/tle/issues/7
```
Comment (juangirini) [inline ts/src/tle.js:27]:
```suggestion
    // https://github.com/ideal-lab5/timelock/issues/7
```

File: ts/src/tle.test.spec.js [hunk 32]
```diff
@@ -0,0 +1,40 @@
+import { describe } from '@jest/globals';
+import { timelockEncrypt, IdealNetworkIdentityHandler } from './tle';
```
Comment (juangirini) [inline ts/src/tle.test.spec.js:2]:
```suggestion
import { timelockEncrypt, IdealNetworkIdentityHandler } from './timelock';
```

File: ts/src/tle.test.spec.js [hunk 33]
```diff
@@ -0,0 +1,40 @@
...
+    // })
+    // it('should decrypt a message on demand if the user knows the secret', async () => {
+    //   const secret = "shhh, it's a secret";
+    //   const ciphertext = 'ciphertext'
+    //   const result = await decrypt(ciphertext, secret);
+    //   expect(result).toEqual({
+    //     message: 'mocked-decrypted',
+    //     sk: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1]
+    //   })
+    // })
```
Comment (juangirini) [inline ts/src/tle.test.spec.js:39]:
```suggestion
```

**Review by juangirini — CHANGES_REQUESTED** [review]
I started reviewing it and got bored of code clean-up comments.
This needs some "search and replace" `tle`-> `timelock`, 'etf` -> `timelock`, `ETF Network` -> `Ideal Network`. File renaming, probably? Removing commented-out code.

**Comment by driemworks** [comment]
Closing in place of simpler PR that only impacts the rust components, will address typescript wrapper in a new PR as well, tracked in https://github.com/ideal-lab5/timelock/issues/9 and https://github.com/ideal-lab5/timelock/issues/6

**Comment by driemworks** [comment]
Closing in place of simpler PR that only impacts the rust components, will address typescript wrapper in a new PR as well, tracked in https://github.com/ideal-lab5/timelock/issues/9 and https://github.com/ideal-lab5/timelock/issues/6

**Pull request closed without merging**